rectools/visuals/visual_app.py

```diff
@@ -0,0 +1,273 @@
+
+    def _check_columns_present_in_recos(self, recos: tp.Dict[tp.Hashable, pd.DataFrame]) -> None:
+        required_columns = {Columns.User, Columns.Item} if self.is_u2i else {Columns.TargetItem, Columns.Item}
+        for model_name, model_recos in recos.items():
+            actual_columns = set(model_recos.columns)
+            if not actual_columns >= required_columns:
+                raise KeyError(f"Missed columns {required_columns - actual_columns} in {model_name} recommendations df")
+
+    def _prepare_interactions_for_i2i(self) -> pd.DataFrame:
+        request_ids = set()
+        for recos_df in self.recos.values():
+            request_ids.update(set(recos_df[Columns.TargetItem].unique()))
+        interactions = pd.DataFrame({Columns.TargetItem: list(request_ids), Columns.Item: list(request_ids)})
+        return interactions
+
+
+class VisualApp:
+    r"""
+    Main tool for recommendations visualization. Creates Jupyter widgets for visual analysis and
+    comparison of recommendations. Outputs both interactions history and recos for all of the
+    requests. `VisualApp` supports user-to-item (u2i) and item-to-item (i2i) cases.
+
+    Models for comparison will be listed from the `recos` dictionary keys.
+    Users (for u2i case) or items (for i2i case) display names for comparison will be listed from the
+    `requests_dict` keys and ids will be taken for values.
+    For u2i case please provide `interactions` and set `is_u2i` = True. For i2i case keep
+    `interactions` as a default `None` and set `is_u2i` = False.
+
+    Parameters
+    ----------
+    recos : tp.Dict[tp.Hashable, pd.DataFrame]
+        Recommendations from different models in a form of a dict. Model names are supposed to be
+        dict keys. Recommendations from models are supposed to be in form of pandas DataFrames with
+        columns:
+            - `Columns.Item` - recommended item id
+            - `Columns.User` - user id (only for u2i)
+            - `Columns.TargetItem` - target item id (only for i2i)
+            - Any other columns that you wish to display in widgets (e.g. rank or score)
+        The original order of the rows will be preserved. Keep in mind to sort the rows correctly
+        before visualizing. The most intuitive wy is to sort by rank in ascending order.
+    item_data : pd.DataFrame
+        Data for items that is used for visualisation in both interactions and recos widgets.
+        Supposed to be in form of a pandas DataFrame with columns:
+            - `Columns.Item` - item id
+            - Any other columns with item data (e.g. name, category, popularity, image link)
+    requests_dict : tp.Dict[tp.Hashable, tp.Hashable]
+        Predefined requests that will be displayed in widgets. For u2i case specific users are
+        considered as requests. For i2i case - specific items. Their names must be specified as keys
+        of the dict and ids as values of the dict.
+    is_u2i : bool, optional
```

Comment:
Please add default value to the line where argument is defined

Reply:
fixed